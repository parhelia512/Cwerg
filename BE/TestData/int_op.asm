# INT_OP


# ========================================
.fun shr_u32 NORMAL [] = [U32 U32 U32]
  .reg U32 [n r val max step]
  .reg U32 [tmp_op1 tmp_op2 tmp_out ]
  .bbl prolog
    poparg val
    poparg max
    poparg step
    mov n 0
  .bbl loop
    shr r val n
    conv tmp_op1 val
    conv tmp_op2 n
    conv tmp_out r

    pusharg tmp_op2
    pusharg tmp_op1
    pusharg tmp_out
    bsr print_x_x_x_ln

    add n n step
    ble  n max loop
    ret


.fun shr_u16 NORMAL [] = [U16 U16 U16]
  .reg U16 [n r val max step]
  .reg U32 [tmp_op1 tmp_op2 tmp_out ]
  .bbl prolog
    poparg val
    poparg max
    poparg step
    mov n 0
  .bbl loop
    shr r val n
    conv tmp_op1 val
    conv tmp_op2 n
    conv tmp_out r

    pusharg tmp_op2
    pusharg tmp_op1
    pusharg tmp_out
    bsr print_x_x_x_ln

    add n n step
    ble  n max loop
    ret

.fun shr_u8 NORMAL [] = [U8 U8 U8]
  .reg U8 [n r val max step]
  .reg U32 [tmp_op1 tmp_op2 tmp_out ]
  .bbl prolog

    poparg val
    poparg max
    poparg step
    mov n 0
  .bbl loop
    shr r val n
    conv tmp_op1 val
    conv tmp_op2 n
    conv tmp_out r

    pusharg tmp_op2
    pusharg tmp_op1
    pusharg tmp_out
    bsr print_x_x_x_ln

    add n n step
    ble  n max loop
    ret

# ========================================
.fun shl_u32 NORMAL [] = [U32 U32 U32]
  .reg U32 [n r val max step]
  .reg U32 [tmp_op1 tmp_op2 tmp_out ]

    .bbl prolog

    poparg val
    poparg max
    poparg step
    mov n 0
  .bbl loop
    shl r val n
    conv tmp_op1 val
    conv tmp_op2 n
    conv tmp_out r
    pusharg tmp_op2
    pusharg tmp_op1
    pusharg tmp_out
    bsr print_x_x_x_ln

    add n n step
    ble  n max loop
    ret


.fun shl_u16 NORMAL [] = [U16 U16 U16]
  .reg U16 [n r val max step]
  .reg U32 [tmp_op1 tmp_op2 tmp_out ]
  .bbl prolog

    poparg val
    poparg max
    poparg step
    mov n 0
  .bbl loop
    shl r val n
    conv tmp_op1 val
    conv tmp_op2 n
    conv tmp_out r
    pusharg tmp_op2
    pusharg tmp_op1
    pusharg tmp_out
    bsr print_x_x_x_ln

    add n n step
    ble  n max loop
    ret

.fun shl_u8 NORMAL [] = [U8 U8 U8]
  .reg U8 [n r val max step]
  .reg U32 [tmp_op1 tmp_op2 tmp_out ]
  .bbl prolog

    poparg val
    poparg max
    poparg step
    mov n 0
  .bbl loop
    shl r val n
    conv tmp_op1 val
    conv tmp_op2 n
    conv tmp_out r
    pusharg tmp_op2
    pusharg tmp_op1
    pusharg tmp_out
    bsr print_x_x_x_ln

    add n n step
    ble  n max loop
    ret

# ========================================
.fun shr_s32 NORMAL [] = [S32 S32 S32]
  .reg S32 [n r val max step]
  .reg U32 [tmp_op1 tmp_op2 tmp_out ]
  .bbl prolog
    poparg val
    poparg max
    poparg step
    mov n 0
  .bbl loop
    shr r val n
    conv tmp_op1 val
    conv tmp_op2 n
    conv tmp_out r
    pusharg tmp_op2
    pusharg tmp_op1
    pusharg tmp_out
    bsr print_x_x_x_ln

    add n n step
    ble  n max loop
    ret


.fun shr_s16 NORMAL [] = [S16 S16 S16]
  .reg S16 [n r val max step]
  .reg U32 [tmp_op1 tmp_op2 tmp_out ]
  .bbl prolog
    poparg val
    poparg max
    poparg step
    mov n 0
  .bbl loop
    shr r val n
    conv tmp_op1 val
    conv tmp_op2 n
    conv tmp_out r
    pusharg tmp_op2
    pusharg tmp_op1
    pusharg tmp_out
    bsr print_x_x_x_ln


    add n n step
    ble  n max loop
    ret

.fun shr_s8 NORMAL [] = [S8 S8 S8]
  .reg S8 [n r val max step]
  .reg U32 [tmp_op1 tmp_op2 tmp_out ]
  .bbl prolog
    poparg val
    poparg max
    poparg step
    mov n 0
  .bbl loop
    shr r val n
    conv tmp_op1 val
    conv tmp_op2 n
    conv tmp_out r
    pusharg tmp_op2
    pusharg tmp_op1
    pusharg tmp_out
    bsr print_x_x_x_ln

    add n n step
    ble  n max loop
    ret


.fun cntlz_u8 NORMAL [] = [U8 U8 U8]
  .reg U8 [r val max step]
  .reg U32 [tmp_op tmp_out ]
  .bbl prolog

    poparg val
    poparg max
    poparg step
  .bbl loop
    cntlz r val
    conv tmp_op val
    conv tmp_out r
    pusharg tmp_op
    pusharg tmp_out
    bsr print_x_x_ln

    add val val step
    ble  val max loop
    ret

.fun cntlz_u32 NORMAL [] = [U32 U32 U32]
  .reg U32 [r val max step]
  .reg U32 [tmp_op tmp_out ]
  .bbl prolog

    poparg val
    poparg max
    poparg step
  .bbl loop
    cntlz r val
    conv tmp_op val
    conv tmp_out r
    pusharg tmp_op
    pusharg tmp_out
    bsr print_x_x_ln

    add val val step
    ble  val max loop
    ret


.fun cnttz_u8 NORMAL [] = [U8 U8 U8]
  .reg U8 [r val max step]
  .reg U32 [tmp_op tmp_out ]
  .bbl prolog

    poparg val
    poparg max
    poparg step
  .bbl loop
    cnttz r val
    conv tmp_op val
    conv tmp_out r
    pusharg tmp_op
    pusharg tmp_out
    bsr print_x_x_ln

    add val val step
    ble  val max loop
    ret

 .fun cnttz_u32 NORMAL [] = [U32 U32 U32]
  .reg U32 [r val max step]
  .reg U32 [tmp_op tmp_out ]
  .bbl prolog

    poparg val
    poparg max
    poparg step
  .bbl loop
    cnttz r val
    conv tmp_op val
    conv tmp_out r
    pusharg tmp_op
    pusharg tmp_out
    bsr print_x_x_ln

    add val val step
    ble  val max loop
    ret

# ========================================
.fun cntpop_u32 NORMAL [] = [U32 U32 U32]
  .reg U32 [r val max step]
  .reg U32 [tmp_op tmp_out ]
  .bbl prolog

    poparg val
    poparg max
    poparg step
  .bbl loop
    cntpop r val
    conv tmp_op val
    conv tmp_out r
    pusharg tmp_op
    pusharg tmp_out
    bsr print_x_x_ln

    add val val step
    ble  val max loop
    ret

# ========================================
.fun main NORMAL [S32] = []
    .reg U32 [ru32 nu32]
    .reg U16 [ru16 nu16]
    .reg U8 [ru8 nu8]
    .reg S32 [rs32 ns32]
    .reg S16 [rs16 ns16]
    .reg S8 [rs8 ns8]

.bbl start
    pusharg 10:U8
    bsr print_c_ln
    pusharg 4:U32
    pusharg 64:U32
    pusharg  0x87654321:U32
    bsr shr_u32

    pusharg 10:U8
    bsr print_c_ln
    pusharg 4:U16
    pusharg 32:U16
    pusharg  0x8765:U16
    bsr shr_u16

    pusharg 10:U8
    bsr print_c_ln
    pusharg 2:U8
    pusharg 16:U8
    pusharg  0x87:U8
    bsr shr_u8

##########
    pusharg 10:U8
    bsr print_c_ln
    pusharg 4:U32
    pusharg 64:U32
    pusharg  0x87654321:U32
    bsr shl_u32

    pusharg 10:U8
    bsr print_c_ln
    pusharg 4:U16
    pusharg 32:U16
    pusharg  0x8765:U16
    bsr shl_u16

    pusharg 10:U8
    bsr print_c_ln
    pusharg 2:U8
    pusharg 16:U8
    pusharg  0x87:U8
    bsr shl_u8

##########
    pusharg 10:U8
    bsr print_c_ln
    pusharg 4:S32
    pusharg 64:S32
    pusharg  -2023406815:S32 # 0x87654321
    bsr shr_s32

    pusharg 10:U8
    bsr print_c_ln
    pusharg 4:S16
    pusharg 32:S16
    pusharg -30875:S16 # 0x8765
    bsr shr_s16

    pusharg 10:U8
    bsr print_c_ln
    pusharg 2:S8
    pusharg 16:S8
    pusharg -121:S8 # 0x87
    bsr shr_s8

##########
    pusharg 10:U8
    bsr print_c_ln
    pusharg 4:S32
    pusharg 64:S32
    pusharg  0x12345678:S32
    bsr shr_s32

    pusharg 10:U8
    bsr print_c_ln
    pusharg 4:S16
    pusharg 32:S16
    pusharg 0x1234:S16
    bsr shr_s16

    pusharg 10:U8
    bsr print_c_ln
    pusharg 2:S8
    pusharg 16:S8
    pusharg 0x12:S8
    bsr shr_s8

##########
    pusharg 10:U8
    bsr print_c_ln
    pusharg 1:U8
    pusharg 16:U8
    pusharg 0x0:U8
    bsr cntlz_u8

    pusharg 10:U8
    bsr print_c_ln
    pusharg 1:U32
    pusharg 16:U32
    pusharg 0x0:U32
    bsr cntlz_u32

##########
    pusharg 10:U8
    bsr print_c_ln
    pusharg 1:U8
    pusharg 16:U8
    pusharg 0x0:U8
    bsr cnttz_u8

    pusharg 10:U8
    bsr print_c_ln
    pusharg 1:U32
    pusharg 16:U32
    pusharg 0x0:U32
    bsr cnttz_u32

##########
    pusharg 10:U8
    bsr print_c_ln
    pusharg 0x10001:U32
    pusharg 0x100010:U32
    pusharg 0x0:U32
    bsr cntpop_u32

##########
    pusharg 0:S32
    ret
