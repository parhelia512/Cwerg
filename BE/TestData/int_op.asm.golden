

87654321 87654321 0
8765432 87654321 4
876543 87654321 8
87654 87654321 c
8765 87654321 10
876 87654321 14
87 87654321 18
8 87654321 1c
87654321 87654321 20
8765432 87654321 24
876543 87654321 28
87654 87654321 2c
8765 87654321 30
876 87654321 34
87 87654321 38
8 87654321 3c
87654321 87654321 40


8765 8765 0
876 8765 4
87 8765 8
8 8765 c
8765 8765 10
876 8765 14
87 8765 18
8 8765 1c
8765 8765 20


87 87 0
21 87 2
8 87 4
2 87 6
87 87 8
21 87 a
8 87 c
2 87 e
87 87 10


87654321 87654321 0
76543210 87654321 4
65432100 87654321 8
54321000 87654321 c
43210000 87654321 10
32100000 87654321 14
21000000 87654321 18
10000000 87654321 1c
87654321 87654321 20
76543210 87654321 24
65432100 87654321 28
54321000 87654321 2c
43210000 87654321 30
32100000 87654321 34
21000000 87654321 38
10000000 87654321 3c
87654321 87654321 40


8765 8765 0
7650 8765 4
6500 8765 8
5000 8765 c
8765 8765 10
7650 8765 14
6500 8765 18
5000 8765 1c
8765 8765 20


87 87 0
1c 87 2
70 87 4
c0 87 6
87 87 8
1c 87 a
70 87 c
c0 87 e
87 87 10


87654321 87654321 0
f8765432 87654321 4
ff876543 87654321 8
fff87654 87654321 c
ffff8765 87654321 10
fffff876 87654321 14
ffffff87 87654321 18
fffffff8 87654321 1c
87654321 87654321 20
f8765432 87654321 24
ff876543 87654321 28
fff87654 87654321 2c
ffff8765 87654321 30
fffff876 87654321 34
ffffff87 87654321 38
fffffff8 87654321 3c
87654321 87654321 40


ffff8765 ffff8765 0
fffff876 ffff8765 4
ffffff87 ffff8765 8
fffffff8 ffff8765 c
ffff8765 ffff8765 10
fffff876 ffff8765 14
ffffff87 ffff8765 18
fffffff8 ffff8765 1c
ffff8765 ffff8765 20


ffffff87 ffffff87 0
ffffffe1 ffffff87 2
fffffff8 ffffff87 4
fffffffe ffffff87 6
ffffff87 ffffff87 8
ffffffe1 ffffff87 a
fffffff8 ffffff87 c
fffffffe ffffff87 e
ffffff87 ffffff87 10


12345678 12345678 0
1234567 12345678 4
123456 12345678 8
12345 12345678 c
1234 12345678 10
123 12345678 14
12 12345678 18
1 12345678 1c
12345678 12345678 20
1234567 12345678 24
123456 12345678 28
12345 12345678 2c
1234 12345678 30
123 12345678 34
12 12345678 38
1 12345678 3c
12345678 12345678 40


1234 1234 0
123 1234 4
12 1234 8
1 1234 c
1234 1234 10
123 1234 14
12 1234 18
1 1234 1c
1234 1234 20


12 12 0
4 12 2
1 12 4
0 12 6
12 12 8
4 12 a
1 12 c
0 12 e
12 12 10


8 0
7 1
6 2
6 3
5 4
5 5
5 6
5 7
4 8
4 9
4 a
4 b
4 c
4 d
4 e
4 f
3 10


20 0
1f 1
1e 2
1e 3
1d 4
1d 5
1d 6
1d 7
1c 8
1c 9
1c a
1c b
1c c
1c d
1c e
1c f
1b 10


8 0
0 1
1 2
0 3
2 4
0 5
1 6
0 7
3 8
0 9
1 a
0 b
2 c
0 d
1 e
0 f
4 10


20 0
0 1
1 2
0 3
2 4
0 5
1 6
0 7
3 8
0 9
1 a
0 b
2 c
0 d
1 e
0 f
4 10


0 0
2 10001
2 20002
4 30003
2 40004
4 50005
4 60006
6 70007
2 80008
4 90009
4 a000a
6 b000b
4 c000c
6 d000d
6 e000e
8 f000f
2 100010
